### PR TITLE
Include Copyright owner/date in HTML license report; show multiple licenses

### PR DESCRIPTION
**Pull request by DonnKey**
This change addresses issue #109 (nee #50) about including Copyright text in the report.
It also handles the rare (but real -- see "Checker Qual") case where there are multiple license kinds by including all the licenses. Some minor changes to improve readability of both the raw html and the final result.

The raw information in the POM is often incomplete in this regard, so it makes a "best guess" when that happens. Those are easily discovered and edited in the raw html if needed.

Incidentally, it also adds another spelling of the MIT license URL (with a .php suffix). I didn't check whether the other "standard" licenses have that form.

### CODE REVIEW

**Comment by jaredsburrows** [comment]
Can you resolve the conflicts?

**Comment by DonnKey** [comment]
I resolved the conflicts. The edit in GitHub dropped a }, which I've fixed locally but not pushed yet.

However... I also realized I hadn't run the regressions, and expected that it would fail because of my changes. I tried to fix that, and ran up against test harness problems. (Several.)

Details follow below.

Before getting into the test harness issues, there are a couple of cosmetic issues that I'm stuck on. I don't like to blame the tools, but in this case I'm stumped and it really does look like the tools to me.  (I've been doing this stuff for a very long time, so that statement is not lightly made, but I also freely admit I could be wrong.)

The resulting HTML has some problems with whitespace/indentations around the `<a>` entries (and a `<br>`, but that might be a cascade); it looks to me as if the code to adjust indentation is simply losing track of where it is.  I've spent several days causing the code to generate "wrong" html trying to determine if there's some alternate pattern I should be following, and the `<a>` simply sticks at the left margin. I'm open to suggestions, but I lean toward living with it.  (The html is correct and as intended except for the whitespace.)

Here's a snippet. Note where the `<a>` entries align.
``` <h3>Notice for packages:</h3>
    <ul>
      <li><a href="#-1748871456">Android SDK Fabric (1.4.8)</a>
        <dl>
          <dt>Copyright &copy; 20xx The original author or authors</dt>
        </dl>
      </li>
<a name="-1748871456"></a>
      <pre>Fabric Software and Services Agreement
<a href="https://fabric.io/terms">https://fabric.io/terms</a></pre>
<br>
      <hr>
```

The problems I've run across in the test harness are not yours, I don't think, but I'm not sure how to hack around them.  (Only the first of the errors is reported at any time, so you have to unmask them one at a time.) Let me know how you'd like to proceed. (My preference is to check the "right answer" in "as intended", and try to fix the test harness over time - but that's messy for you!)

1. Fatal error in assertHtml because the resulting HTML contains `&copy;` : it's being reported as an "undefined entity" even if it is official HTML. (Yeah, using `(c)` would work, but the copyright symbol is available and a better choice.)
2. Fatal error in assertHtml because there's no matching `</hr>` for the `<hr>`. Except that `<hr>` doesn't require a matching `</hr>`! (Some HTML dialects apparently do, but at least w3schools says `<hr>` has no end tag for straight HTML.) The `<hr>` really makes the result more readable, and I haven't a workaround in mind.
3. Fatal error like the above for `<br>`. Here there's no ambiguity: `<br>` has never had an end tag, and it makes no sense to have one.  The `<br>` is really needed, and I don't see an alternative.
4. I'm using Windows, and it looks as if there's a problem (another one) with respect to Windows path names. I'm seeing the following
```
result.output.find("Wrote HTML report to .*${reportFolder}/${taskName}.html.")
|      |      |                             |               |
|      |      |                             |               licenseDebugReport
|      |      |                             C:\Users\donnt\AppData\Local\Temp\junit2089520722103704024/build/reports/licenses
|      |      java.util.regex.PatternSyntaxException: Illegal/unsupported escape sequence near index 26
|      |      Wrote HTML report to .*C:\Users\donnt\AppData\Local\Temp\junit2089520722103704024/build/reports/licenses/licenseDebugReport.html.
|      |                                ^
|       
```
Note the caret under the \U - it looks as if it's trying to treat the Windows path separator as an escape.  I'm inclined to fix everything else however we decide to do it, and rely on the automated testing to deal with that (give up testing on Windows).  (Yeah, I know, "Windows"...).

Lastly, for the record if you haven't seen it, I'm seeing a gripe about Groovy internals. Presumably someone is working on it, but for the record:

* An illegal reflective access operation has occurred 
* Illegal reflective access by org.codehaus.groovy.vmplugin.v7.Java7$1 (file:/C:/Users/donnt/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1/lib/groovy-all-1.0-2.5.4.jar) to constructor java.lang.invoke.MethodHandles$Lookup(java.lang.Class,int) 
* Please consider reporting this to the maintainers of org.codehaus.groovy.vmplugin.v7.Java7$1 
* Use --illegal-access=warn to enable warnings of further illegal reflective access operations All illegal access operations will be denied in a future release

**Comment by DonnKey** [comment]
Over night it occurred to me to look at exactly what assertHtml did (I had assumed it was a built-in). The problem is that it's comparing XML when the input is HTML, and it matters in this case. I didn't see any HTML comparison tools similar to DiffBuilder, and I suspect that `<hr>` and `<br>` can't be dealt with at all because it's assuming well-formed (properly closed) XML.  Either a preliminary translation phase to convert the HTML into legal XML (e.g. `<hr>` -> `<hr/>`) or a completely different comparison strategy will be needed. As final owner (and arbiter) I think at least the outline of the solution should be your call.

**Comment by jaredsburrows** [comment]
@DonnKey If you have a better library or way to compare the HTML, please let me know.

**Comment by DonnKey** [comment]
I don't know of anything that would work directly, and didn't find anything on the web that looked at all promising.  Pre-transforming the strings to make them into something the XML comparison will accept seems possible (subject to actually trying it). Of course, straight string comparison would work but be less resilient to unintended changes (such as the indentation of `<a>` elements). If you want me to give pre-transformation a try, I can, if it would be something you'd accept if it worked.

File: src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt [hunk 1]
```diff
@@ -1,24 +1,10 @@
...
-import kotlinx.html.HtmlTagMarker
-import kotlinx.html.a
-import kotlinx.html.attributesMapOf
-import kotlinx.html.body
-import kotlinx.html.h3
-import kotlinx.html.head
-import kotlinx.html.html
-import kotlinx.html.li
-import kotlinx.html.pre
+import kotlinx.html.*
```
Comment (jaredsburrows) [inline src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt]: Can you import everything individually?

File: src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt [hunk 2]
```diff
@@ -42,23 +32,25 @@ class HtmlReport(private val projects: List<Project>) {
     val projectsMap = hashMapOf<String?, List<Project>>()
     val licenseMap = LicenseHelper.licenseMap
 
-    // Store packages by license
+    // Store packages by licenses: build a composite key of all the licenses, sorted in the (probably vain)
+    // hope that there's more than one project with the same set of multiple licenses.
     projects.forEach { project ->
-      var key: String? = ""
+      val keys: MutableList<String> = mutableListOf()
```
Comment (jaredsburrows) [inline src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt]: Can you make this:

val keys = mutableListOf<String>()

File: src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt [hunk 3]
```diff
@@ -42,23 +32,25 @@ class HtmlReport(private val projects: List<Project>) {
...
-          // otherwise, use the url as a key
-          else -> license.url
+        project.licenses.forEach { license ->
+          keys.add(getLicenseKey(license))
         }
       }
 
+      keys.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it })
+      var key = ""
+      if (keys.size > 0) {
```
Comment (jaredsburrows) [inline src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt]: Can you make this: `if (!keys.isEmpty()) {`

File: src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt [hunk 4]
```diff
@@ -156,8 +186,23 @@ class HtmlReport(private val projects: List<Project>) {
...
+      LicenseHelper.licenseMap.containsKey(license.name) -> LicenseHelper.licenseMap[license.name]
+      // otherwise, use the url as a key
+      else -> license.url
+    } as String
+  }
+}
+
+fun getLicenseText(fileName: String): String {
+  val resource = HtmlReport::class.java.getResource("/license/$fileName")
+  return (resource?.readText() ?: HtmlReport.MISSING_LICENSE + fileName)
```
Comment (jaredsburrows) [inline src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt]: Can you remove the parenthesis?

File: src/test/groovy/test/TestUtils.groovy [hunk 5]
```diff
@@ -19,16 +23,29 @@ final class TestUtils {
       .withPluginClasspath()
       .build()
   }
-
-  static def getLicenseText(def fileName) {
-    getClass().getResource("/license/${fileName}").text
+  static def myGetLicenseText(String fileName) {
```
Comment (jaredsburrows) [inline src/test/groovy/test/TestUtils.groovy]: Can you keep this the same name?

Comment (DonnKey) [reply src/test/groovy/test/TestUtils.groovy]: I have no objection to a spelling change (it is rather unimaginative!), but I'm not sure it's wise to end up with two functions with the same name in this case, because the scoping in groovy doesn't seem to be as clear as in Java. I had troubles getting this right (possibly my own fault) until I separated the spellings, and this strikes me as a case where future maintainers could easily find themselves confused.  If you insist..., but it was not an idle decision to change the name.

Comment (jaredsburrows) [reply src/test/groovy/test/TestUtils.groovy]: Sure, they can be different names.

**Comment by jaredsburrows** [comment]
If we are updating the output, should we also update the README.md?

**Comment by DonnKey** [comment]
README is mostly done. I left the indentation problems with the HTML as is. Let me know (either way, please) if you'd prefer that to be corrected in the README for readability reasons.

The CHANGELOG doesn't seem to need anything from me.

File: src/test/groovy/test/TestUtils.groovy (hunk 5 above)
Comment (jaredsburrows) [reply src/test/groovy/test/TestUtils.groovy]: Its ok to change the name but "my" getter is not a normal convention. Can we name this something more reasonable?

**Comment by jaredsburrows** [comment]
@DonnKey It would be nice to keep the README.md up to date with the code. This way people know what they are getting when they use the plugin.

File: src/main/kotlin/com/jaredsburrows/license/internal/LicenseHelper.kt [hunk 6]
```diff
@@ -80,6 +80,7 @@ object LicenseHelper {
     "MIT License" to "mit.txt",
     "http://opensource.org/licenses/MIT" to "mit.txt",
     "https://opensource.org/licenses/MIT" to "mit.txt",
+    "http://www.opensource.org/licenses/mit-license.php" to "mit.txt",
```
Comment (jaredsburrows) [inline src/main/kotlin/com/jaredsburrows/license/internal/LicenseHelper.kt]: Should this be in a separate PR?

File: src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt [hunk 7]
```diff
@@ -32,6 +38,10 @@ class HtmlReport(private val projects: List<Project>) {
     const val NO_LIBRARIES = "None"
     const val NO_LICENSE = "No license found"
     const val NOTICE_LIBRARIES = "Notice for packages:"
+    const val COPYRIGHT = "Copyright "
+    const val DEFAULT_AUTHOR = "The original author or authors"
+    const val DEFAULT_YEAR = "20xx"
```
Comment (jaredsburrows) [inline src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt]: Will this show up if no year is specified?

File: src/main/kotlin/com/jaredsburrows/license/internal/LicenseHelper.kt (hunk 6 above)
Comment (DonnKey) [reply src/main/kotlin/com/jaredsburrows/license/internal/LicenseHelper.kt]: If you'd prefer. I'll back it out for now.

File: src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt (hunk 7 above)
Comment (DonnKey) [reply src/main/kotlin/com/jaredsburrows/license/internal/report/HtmlReport.kt]: Lines 42 and 43 are default text in case the information is absent in the POM. My thought was to leave enough structure there that it's easy to hand-edit any corrections if the user wants to go to that effort. Removal of the whole thing is easily done if that's the preference.

I'll push an update shortly.

File: src/test/groovy/test/TestUtils.groovy [hunk 8]
```diff
@@ -19,16 +23,29 @@ final class TestUtils {
...
     return parser.parse(actual) == parser.parse(expected)
   }
 
+  static def htmlToXml(String text) {
+    // Convert HTML into legal-enough XML that we can use the XML comparison
+    // utility to compare two HTML strings. This is only just what we need for
+    // this exact case, so update as needed.
+    text = text.replaceAll('<br>', '<br/>')
+    text = text.replaceAll('<hr>', '<hr/>')
+    text = text.replaceAll('&copy;', '(c)')
```
Comment (jaredsburrows) [inline src/test/groovy/test/TestUtils.groovy]: Sorry, one last thing before we merge, why remove copy with `(c)`?

File: src/test/groovy/com/jaredsburrows/license/LicensePluginJavaSpec.groovy [hunk 9]
```diff
@@ -724,23 +744,26 @@ final class LicensePluginJavaSpec extends Specification {
...
-          </style>
+          <style>body { font-family: sans-serif } pre { background-color: #eeeeee; padding: 1em; white-space: pre-wrap; display: inline-block }</style>
           <title>Open source licenses</title>
         </head>
         <body>
           <h3>Notice for packages:</h3>
           <ul>
-            <li>
-              <a href="#314129783">appcompat-v7</a>
+            <li><a href="#1934118923">appcompat-v7 (26.1.0)</a>
```
Comment (jaredsburrows) [inline src/test/groovy/com/jaredsburrows/license/LicensePluginJavaSpec.groovy]: Sorry, one last thing before we merge, can we remove the parentheses around the version?

File: src/test/groovy/test/TestUtils.groovy (hunk 8 above)
Comment (DonnKey) [reply src/test/groovy/test/TestUtils.groovy]: The XML comparison recognizes that '&copy;' is an "entity" (ampersand-semicolon), but doesn't have the copyright entity built-in, yielding a complaint about "undefined entity" and failing the test.  The replacement doesn't really matter, but that seemed obvious.  (The message the XML comparison generates is obvious only after you've figured out what it means!)

File: src/test/groovy/com/jaredsburrows/license/LicensePluginJavaSpec.groovy (hunk 9 above)
Comment (DonnKey) [reply src/test/groovy/com/jaredsburrows/license/LicensePluginJavaSpec.groovy]: I did that explicitly so that if there wasn't a version (theoretically possible, but, granted, unlikely) there's be an obvious placeholder that could be edited. I don't mind making the change, but there was a reason for doing it this way. If you still want a change, would you prefer no placeholder or some other placeholder?

**Comment by jaredsburrows** [comment]
Thanks!
